logins-api/android/library/src/main/java/org/mozilla/sync15/logins/rust/PasswordSyncAdapter.kt

```diff
@@ -53,7 +53,9 @@ internal interface PasswordSyncAdapter : Library {
     fun sync15_passwords_reset(state: RawLoginSyncState, error: RustError.ByReference)
 
     fun sync15_passwords_touch(state: RawLoginSyncState, id: String, error: RustError.ByReference)
-    fun sync15_passwords_delete(state: RawLoginSyncState, id: String, error: RustError.ByReference): Boolean
+    // This is 1 for true and 0 for false, it would be a boolean but we need to return a value with
```

Comment:
Do you have any thoughts on how we can make this better for producers?  It's awkward to have to do manual serialization at the boundaries at all times.

Reply:
There are three possible solutions

- Rust type trickery
- Linting
- Code generation

The first misses a large class of errors, which is where the a declaration on the Kotlin FFI is done subtly incorrectly (usually [using String instead of Pointer](https://github.com/mozilla-mobile/android-components/pull/710), although the same issue happened a few times during development). Even for just rust code there are issues around unclear types being fundamentally a big risk for FFI code that depends on the native bit-level representation of those types...

A linter that understands both Kotlin and Rust would be one option that hits both. Not sure how much work this is.

For codegen it's probably a lot of work, but not clear exactly how much. I played around with wasm_bindgen recently and it gave me some ideas, although it's unclear how monumental an effort it would be to do something like fork them and make it work for something like this. That said, what they have on offer is extremely compelling as an example showing "yes this can be done" (https://github.com/rustwasm/wasm-bindgen/blob/master/examples/smorgasboard/src/lib.rs#L20-L43 and https://github.com/rustwasm/wasm-bindgen/blob/master/examples/smorgasboard/app.js#L20-L22 on the other end).

Obviously we could also do something hackier than that too for codegen.

Reply:
We'd need to lint the rust side too -- the linter has to understand the API exposed by rust and by kotlin. Bleh. (Obviously in practice it can be two linters that talk though)